### DISCUSSION TITLE
[QUESTION] - Looking for some better way to encode or decode json.

### DISCUSSION THREAD

**Discussion by sujit-baniya** [http4s/http4s#6995] [Q&A]
In Scala 2.11.7 and http4s 0.8.6 (with argonaut),

The following code worked without issue for `AssignCoder` or `Option[AssignCoder]` or `List[AssignCoder]`

```scala

import argonaut.Argonaut._
import argonaut._
import org.http4s.argonaut._

case class AssignCoder(coder: String)

object AssignCoder {
  implicit val codec = CodecJson.casecodec1(AssignCoder.apply, AssignCoder.unapply)("coder")
  //implicit def encode = codec.Encoder
  implicit val entityEncoder = jsonEncoderOf[AssignCoder]
}
```

But now in Scala 2.12.17 and http4s 0.15.16 (with argonaut),

I've to define explicitly for cases as below

```scala

import argonaut.Argonaut._
import argonaut._
import org.http4s.argonaut._

case class AssignCoder(coder: String)

object AssignCoder {
  implicit val codec = CodecJson.casecodec1(AssignCoder.apply, AssignCoder.unapply)("coder")
  //implicit def encode = codec.Encoder
  implicit val entityEncoder = jsonEncoderOf[AssignCoder]
  implicit val optionEntityEncoder = jsonEncoderOf[Option[AssignCoder]]
  implicit val listEntityEncoder = jsonEncoderOf[List[AssignCoder]]
  implicit val listOptionEntityEncoder = jsonEncoderOf[List[Option[AssignCoder]]  
}
```

Any better approach as the above is a bit tricky to add all possible combinations.

**Comment by rossabaker** [comment]
I can't recall any reason that would have changed, but it has been several years since I looked at Argonaut.  I would consider diffing the argonaut source code between those two http4s versions to see whether any implicits changed, and also look at which versions of argonaut were used to see whether its derivation changed.

**Reply by sujit-baniya** [reply]
@rossabaker I did compare for EncoderJson and other instances between the projects, I didn't find/locate the difference

**Reply by rossabaker** [reply]
This is a longshot, but does it work better if you put explicit types on `codec` and `entityEncoder`?

**Comment by MazieRamyqfW** [comment]
Instead of defining implicit entity encoders for each case class and its variations, you could define a generic encoder for any case class that has a corresponding CodecJson. Here's an example implementation:

import argonaut.Argonaut._
import argonaut._

object JsonCodecs {
  implicit def genericEntityEncoder[A: CodecJson]: EntityEncoder[IO, A] =
    jsonEncoderOf[IO, A]
}

Then, you can import the genericEntityEncoder and use it for any case class that has a corresponding CodecJson, without having to define additional implicit entity encoders:

import org.http4s.EntityEncoder
import org.http4s.argonaut.jsonEncoderOf
import org.http4s.dsl.io._
import org.http4s.implicits._
import JsonCodecs._

case class AssignCoder(coder: String)

object AssignCoder {
  implicit val codec = CodecJson.casecodec1(AssignCoder.apply, AssignCoder.unapply)("coder")
}

val assignCoder = AssignCoder("foo")

// Encode single instance
val response = Ok(assignCoder)

// Encode Option[AssignCoder]
val response = Ok(Option(assignCoder))

// Encode List[AssignCoder]
val response = Ok(List(assignCoder, assignCoder))

This approach should work for any case class that has a corresponding CodecJson, without having to define additional implicit entity encoders.
